### PR TITLE
Django move from heroku to railway

### LINKED ISSUE #19987: Comment on the example for installing express on Heroku: Heroku will phase out free plans for dynos, and shut down free dynos from November 28, 2022
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/deployment

### What specific section or headline is this issue about?

Example: Installing LocalLibrary on Heroku

### What information was incorrect, unhelpful, or incomplete?

> Heroku has a [free tier](https://www.heroku.com/pricing) that is really free (albeit with some limitations).

### What did you expect to see?

I stumbled upon this as I was looking for an alternative to Heroku after learning that they will be phasing out the said tier.

### Do you have any supporting links, references, or citations?

As per heroku's [blog](https://blog.heroku.com/next-chapter), they will be phasing out free plans in heroku's dynos. Here is the statement quoted from the Heroku's blog

> Our product, engineering, and security teams are spending an extraordinary amount of effort to manage fraud and abuse of the Heroku free product plans. In order to focus our resources on delivering mission-critical capabilities for customers, we will be phasing out our free plan for Heroku Dynos, free plan for Heroku Postgres, and free plan for Heroku Data for Redis®, as well as deleting inactive accounts.

>Starting October 26, 2022, we will begin deleting inactive accounts and associated storage for accounts that have been inactive for over a year. Starting November 28, 2022, we plan to stop offering free product plans and plan to start shutting down free dynos and data services. We will be sending out a series of email communications to affected users.

>We will continue to provide low-cost solutions for compute and data resources: Heroku Dynos starts at $7/month, Heroku Data for Redis® starts at $15/month, Heroku Postgres starts at $9/month. See [Heroku Pricing Information](https://www.heroku.com/pricing) for current details. These include all the features of the free plans with additional certificate management and the assurance your dynos do not sleep to help ensure your apps are responsive.

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/server-side/express_nodejs/deployment`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/deployment
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/server-side/express_nodejs/deployment/index.md
* Last commit: https://github.com/mdn/content/commit/51f2cde58fff4a44242577b2540bdfdfc525dda4
* Document last modified: 2022-08-13T04:30:09.000Z

</details>

### PR DESCRIPTION
**Pull request by hamishwillee**
Heroku no longer has free plans. MDN team evaluation seems to think Railway is a good option. This is docs update work.

Fixes #19987

Associated update to tutorial: https://github.com/mdn/django-locallibrary-tutorial/pull/109 (needs one more minor change)

### CODE REVIEW

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 1]
```diff
@@ -660,7 +744,16 @@ The next step is to read our last few articles, and then complete the assessment
...
+- Railway Docs
+
+  - [CLI](https://docs.railway.app/develop/cli)
+
+- Digital Ocean
+
+  - [How To Serve Django Applications with uWSGI and Nginx on Ubuntu 16.04](https://www.digitalocean.com/community/tutorials/how-to-serve-django-applications-with-uwsgi-and-nginx-on-ubuntu-16-04)
+  - [Other Digital Ocean Django community docs](https://www.digitalocean.com/community/tutorials?q=django)
+
+- Heroku Docs (similar setup concepts)
```
Comment (hamishwillee) [inline files/en-us/learn/server-side/django/deployment/index.md:756]: FYI only, was tempting to delete these, but the Railway docs are far more limited, and many of the concepts in these are useful. So I have included them as "for interest" near the bottom.

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 2]
```diff
@@ -147,91 +152,105 @@ A full checklist of settings you might want to change is provided in [Deployment
...
+Once the application is running it can configure itself using information provided in [environment variables](https://docs.railway.app/develop/variables).
+For example, an application that uses a database can get the address using the variable `DATABASE_URL`.
+The database service itself may be hosted by Railway or some other provider.
 
-In order to execute your application Heroku needs to be able to set up the appropriate environment and dependencies, and also understand how it is launched. For Django apps we provide this information in a number of text files:
+Developers interact with Railway through the Railway site, and using a special [Command Line Interface (CLI)](https://docs.railway.app/develop/cli) tool.
+The CLI allows you to associate a local Github repository with a railway project, upload the repository from the local branch to the live site, inspect the logs of the running process, set and get configuration variables and much more.
+One of the most useful features is you can use the CLI to run your local project with the same environment variables as the live project.
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:204]:
```suggestion
One of the most useful features is that you can use the CLI to run your local project with the same environment variables as the live project.
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 3]
```diff
@@ -147,91 +152,105 @@ A full checklist of settings you might want to change is provided in [Deployment
...
 
-- **runtime.txt**: the programming language and version to use.
-- **requirements.txt**: the Python component dependencies, including Django.
-- **Procfile**: A list of processes to be executed to start the web application. For Django this will usually be the Gunicorn web application server (with a `.wsgi` script).
-- **wsgi.py**: [WSGI](https://wsgi.readthedocs.io/en/latest/what.html) configuration to call our Django application in the Heroku environment.
+In order to get our application to work on Railway we'll need to put our Django web application into a git repository, add the files above, integrate with a database add-on, and make changes to properly handle static files.
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:206]:
```suggestion
In order to get our application to work on Railway, we'll need to put our Django web application into a git repository, add the files above, integrate with a database add-on, and make changes to properly handle static files.
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 4]
```diff
@@ -147,91 +152,105 @@ A full checklist of settings you might want to change is provided in [Deployment
...
+Once we've done all that we can set up a Railway account, get the Railway client, and install our website.
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:207]:
```suggestion
Once we've done all that, we can set up a Railway account, get the Railway client, and install our website.
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 5]
```diff
@@ -147,91 +152,105 @@ A full checklist of settings you might want to change is provided in [Deployment
...
 
 ### Creating an application repository in GitHub
 
-Heroku is closely integrated with the **git** source code version control system, using it to upload/synchronize any changes you make to the live system. It does this by adding a new heroku "remote" repository named _heroku_ pointing to a repository for your source on the Heroku cloud. During development you use git to store changes on your own repository. When you want to deploy your site, you sync your changes to the Heroku repository.
+Railway is closely integrated with Github and the **git** source code version control system, and you can configure it to automatically deploy changes to a particular repository or branch on Github.
+Alternatively you can push your current local code branch direct to the railway deployment using the CLI.
 
-> **Note:** If you're used to following good software development practices you are probably already using git or some other SCM system. If you already have a git repository, then you can skip this step.
+> **Note:** Using a source code management system like Github is good software development practice.
+> Skip this step if you're already using Github to manager your source.
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:217]:
```suggestion
> Skip this step if you're already using Github to manage your source.
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 6]
```diff
@@ -336,49 +372,63 @@ pip3 install dj-database-url
...
+Static files can therefore be uploaded to any host and you can update your application to find them using this setting.
+
+The _collectstatic_ tool is used to collect static files into the folder defined by the `STATIC_ROOT` project setting.
+It is called with the following command:
+
+```bash
+python3 manage.py collectstatic
+```
+
+For this tutorial _collectstatic_ is run automatically by Railway before the application is uploaded, copying all the static files in the application to the location specified in `STATIC_ROOT`.
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:421]:
```suggestion
For this tutorial, _collectstatic_ is run automatically by Railway before the application is uploaded, copying all the static files in the application to the location specified in `STATIC_ROOT`.
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 7]
```diff
@@ -336,49 +372,63 @@ pip3 install dj-database-url
...
+`Whitenoise` then finds the files from the location defined by `STATIC_ROOT` (by default) and serves them at the base URL defined by `STATIC_URL`.
+
+> **Note:** Railway does not document that it calls `python3 manage.py collectstatic` when a Django application is loaded — this has been verified by testing.
+> If needed you could run the tool by adding the command to the [Procfile](#procfile).
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:425]:
```suggestion
> If needed, you could run the tool by adding the command to the [Procfile](#procfile).
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 8]
```diff
@@ -456,193 +509,244 @@ wheel==0.37.1
...
+![Railway website screen - deployment](railway_project_deploy.png)
 
-If we're lucky, the app is now "running" on the site, but it won't be working properly because we haven't set up the database tables for use by our application. To do this we need to use the `heroku run` command and start a "[one off dyno](https://devcenter.heroku.com/articles/deploying-python#one-off-dynos)" to perform a migrate operation. Enter the following command in your terminal:
+You can click the site URL (highlighted above) to open the site in a browser (it still won't work, because the setup is not complete).
 
-```bash
-heroku run python manage.py migrate
+### Set ALLOWED_HOSTS and CSRF_TRUSTED_ORIGINS
+
+When the site is opened at this point you'll see an error debug screen as shown below.
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:582]:
```suggestion
When the site is opened, at this point you'll see an error debug screen as shown below.
```

File: files/en-us/learn/server-side/django/deployment/index.md [hunk 9]
```diff
@@ -456,193 +509,244 @@ wheel==0.37.1
...
-> heroku config:set DJANGO_DEBUG=False
+railway help
+```
+
+> **Note:** In the following section we use `railway login` and `railway link` to link the current project to a directory.
+> If you are logged out by the system, you will need to call both commands again to re-link the project.
+
+### Configure a superuser
+
+In order to create a superuser we need to call the Django `createsuperuser` command against the production database (this is the same operation as we ran locally in [Django Tutorial Part 4: Django admin site > Creating a superuser](/en-US/docs/Learn/Server-side/Django/Admin_site#creating_a_superuser)).
```
Comment (sideshowbarker) [inline files/en-us/learn/server-side/django/deployment/index.md:687]:
```suggestion
In order to create a superuser, we need to call the Django `createsuperuser` command against the production database (this is the same operation as we ran locally in [Django Tutorial Part 4: Django admin site > Creating a superuser](/en-US/docs/Learn/Server-side/Django/Admin_site#creating_a_superuser)).
```

**Review by sideshowbarker — APPROVED** [review]

**Comment by sideshowbarker** [comment]
Beautiful.

**Comment by hamishwillee** [comment]
Thanks for the review. Now the joy of doing the same thing for Express :-)

**Pull request merged**